modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

```diff
@@ -1255,7 +1258,10 @@ private[smithy4s] class SchemaVisitorJCodec(
       }
 
       def decodeValue(cursor: Cursor, in: JsonReader): Option[A] =
-        if (in.isNextToken('n'))
+        // if `A` is an option and has nullable, we delegate the handling of `null` to it.
+        // This allows for supporting Json-merge patches, where the absence of value
+        // and the presence of "null" have different meanings.
+        if (in.isNextToken('n') && !aIsNullable)
```

Comment:
shouldn't it be `if (in.isNextToken('n') && aIsNullable)` ? my understanding is that if the next token is a `n` AND we are looking for `null`, then we can read a `null`

Reply:
but the test right below tells me everything is good

Reply:
It's testing if the underlying schema is nullable, which means the option can delegate to the underlying decoder for null-handling. If the `A` is not nullable however, the option needs to take care of it.